packages/jsii-pacmak/lib/targets/python/type-name.ts

```diff
@@ -0,0 +1,354 @@
+    }
+    return { [requiredImport.sourcePackage]: new Set([requiredImport.item]) };
+  }
+
+  private resolve({ assembly, submodule, nestingScope, typeAnnotation = true }: NamingContext) {
+    const { assemblyName, pythonFqn } = toPythonFqn(this.#fqn, assembly);
+    if (assemblyName !== assembly.name) {
+      return {
+        pythonType: pythonFqn,
+        requiredImport: {
+          // TODO: Migth we need to import a subpackage instead?
+          sourcePackage: toPythonFqn(assemblyName, assembly).pythonFqn,
+          item: '',
+        },
+      };
+    }
+
+    const submodulePythonName = toPythonFqn(submodule, assembly).pythonFqn;
+    const typeSubmodulePythonName = toPythonFqn(findParentSubmodule(assembly.types![this.#fqn], assembly), assembly).pythonFqn;
+
+    if (typeSubmodulePythonName === submodulePythonName) {
+      // Submodule-local type, so we'll just drop the submodule name prefix here, unless we are
+      // referencing a type within the current nesting context, where we'll want to make a context
+      // local reference by dropping the nesting type's name prefix.
+      const nestingParent = nestingScope && toPythonFqn(nestingScope, assembly).pythonFqn;
+      const localName = nestingParent && pythonFqn.startsWith(`${nestingParent}.`)
+        ? pythonFqn.substring(nestingParent.length + 1)
+        : pythonFqn.substring(typeSubmodulePythonName.length + 1);
+      return {
+        pythonType: typeAnnotation ? JSON.stringify(localName) : localName,
+      };
+    }
+
+    const [toImport, ...nested] = pythonFqn.substring(typeSubmodulePythonName.length + 1).split('.');
+    const aliasSuffix = createHash('sha256')
+      .update(typeSubmodulePythonName)
+      .update('.')
+      .update(toImport)
+      .digest('hex')
+      .substring(0, 8);
+    const alias = `_${toImport}_${aliasSuffix}`;
+
+    return {
+      pythonType: [alias, ...nested].join('.'),
+      requiredImport: {
+        sourcePackage: relativeImportPath(submodulePythonName, typeSubmodulePythonName),
+        item: `${toImport} as ${alias}`,
+      },
+    };
+  }
+}
+
+function toPythonFqn(fqn: string, rootAssm: Assembly) {
+  const [assemblyName, ...qualifiedIdentifiers] = fqn.split('.');
+  const fqnParts: string[] = [moduleName(assemblyName, rootAssm)];
+
+  for (const part of qualifiedIdentifiers) {
+    fqnParts.push(toPythonIdentifier(part));
+  }
+
+  return { assemblyName, pythonFqn: fqnParts.join('.') };
+}
+
+/**
+ * Computes the python relative import path from `fromModule` to `toModule`.
+ *
+ * @param fromPkg the package where the relative import statement is located.
+ * @param toPkg   the package that needs to be relatively imported.
+ *
+ * @returns a relative import path.
+ *
+ * @example
+ *  relativeImportPath('A.B.C.D', 'A.B.E') === '...E';
+ *  relativeImportPath('A.B.C', 'A.B')     === '..';
+ *  relativeImportPath('A.B', 'A.B.C')     === '.C';
+ */
+function relativeImportPath(fromPkg: string, toPkg: string): string {
```

Comment:
Would it make sense instead of these aliases, always import from the root? See example in a comment below.

Reply:
There's one critical case where that fails (see my answer to your comment below). I'd rather have only one way to manage those than have to distinguish "from root" versus "from non-root".